tensorflow/compiler/jit/deadness_analysis.cc

```diff
@@ -1069,6 +1240,52 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
         absl::StrJoin(unreachable_nodes, ", "));
   }
 
+  std::vector<Node*> tpo;
+  TF_RETURN_IF_ERROR(
+      GetFrameBasedTopologicalOrder(&graph_, control_flow_info_, &tpo));
+
+  size_t frame_start = 0;
+  for (size_t i = 0; i < tpo.size(); ++i) {
+    // Collect nodes until we see a node who has a different root frame.
+    if (i != tpo.size() - 1) {
+      string i_frame_name, next_frame_name;
+      TF_RETURN_IF_ERROR(GetRootFrame(tpo[i], control_flow_info_,
+                                      graph_.source_node(), &i_frame_name));
+      TF_RETURN_IF_ERROR(GetRootFrame(tpo[i + 1], control_flow_info_,
+                                      graph_.source_node(), &next_frame_name));
+      if (i_frame_name == next_frame_name) {
+        continue;
+      }
+    }
+
+    string frame_name = control_flow_info_[tpo[i]->id()].frame_name;
+    absl::Span<Node*> sub_tpo(tpo.data() + frame_start, i - frame_start + 1);
+    frame_start = i + 1;
+
+    // First, try the optimistic mode.
+    bool is_success = false;
+    if (!force_pessimistic && !frame_name.empty()) {
+      TF_RETURN_IF_ERROR(
+          PopulateFrame(sub_tpo, /*use_optimistic_mode*/ true, &is_success));
+    }
+    if (!is_success) {
+      // The optimistic mode does not converge. Let's fall back to the
+      // pessimistic mode.
+      TF_RETURN_IF_ERROR(
+          PopulateFrame(sub_tpo, /*use_optimistic_mode*/ false, nullptr));
+    }
+    if (VLOG_IS_ON(2)) {
```

Comment:
This is not necessary -- `VLOG(2)` automatically checks this.  And it has some C++ macro cleverness to avoid evaluating the expressions being streamed into it so checking `VLOG_IS_ON(2)` in this case has no performance benefit either.